### PR TITLE
Make group prefix optional

### PR DESCRIPTION
**Pull request by gregturn**
Make the prefix optional.

### CODE REVIEW

**Comment by gregturn** [comment]
@markfisher @markpollack @sabbyanandan @ericbottard Please provide feedback on this proposal.

I know we need test cases to cover this, but I wanted to run this idea by before going too far.

**Comment by markpollack** [comment]
Hi.  I presume this is because you want to name things in a particular way when deploying so it is managed by spinnaker?  Though, for that I would think the suffix style `_123` would be the important point.

We really even want this to be required (but ATM the base class tests don't support this) since if we don't have something relatively unique, you can't deploy more than one "time | log" stream across the entire install of CF, which is too limiting.  Prefixing the stream name is minimal to avoid this type of cross talk.  I believe there is an issue (can't remember where) to also put a unique name prefix before the stream name that is unique to the dataflow instance.  @sabbyanandan where is that issue?

**Comment by markpollack** [comment]
What is the exact problem you are trying to work around?

**Comment by gregturn** [comment]
With no supplied environment variables, deploying 'clouddriver' causes the application name on PCF to become 'null-clouddriver'. I tweaked the Spinnaker deployer app to plugin a setting of 'default', but that is suboptimal.

The following code insists on this structure...
```
private String deploymentId(AppDeploymentRequest request) {
	return String.format("%s-%s",
		request.getEnvironmentProperties().get(GROUP_PROPERTY_KEY),
		request.getDefinition().getName());
}
```

I want the option to drop the "<group>-" prefix and simply use the name supplied by the caller.

**Comment by gregturn** [comment]
I have now coded a couple unit tests to prove the results with/without an AppDeployer.GROUP_PROPERTY_KEY included in the environment properties.

**Comment by gregturn** [comment]
@markpollack I've rewritten the Spock test as JUnit + Mockito, if you want to approve it.

File: src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryAppDeployer.java [hunk 1]
```diff
@@ -15,14 +15,6 @@
  */
 package org.springframework.cloud.deployer.spi.cloudfoundry;
 
-import static java.lang.Integer.parseInt;
```
Comment (markfisher) [inline src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryAppDeployer.java]: can you please update your import organization settings?... the preexisting code was following the conventions (statics first, empty line before org.springframework, etc)

**Comment by gregturn** [comment]
@markfisher I created another formatter in my IDE and polished up all the import statements. If this is good, I'll squash and merge to master.

**Comment by dsyer** [comment]
I understand the dataflow use case, but I don't think dataflow should be dictating the features of the API, or at least not at this level. The example Greg gave shows how ridiculous this is: you deploy an app called "foo" and Cloud Foundry gets an app called "null-foo", because you didn't know that the (typeless, optional) properties had to contain a special key.

Even if you did know about the key the fact that you can't undeploy an app that you didn't deploy is unnecessarily restrictive IMO (the deploy and undeploy keys are not the same).

**Comment by markfisher** [comment]
@dsyer We definitely do not want dataflow to dictate the features of this API, but I think "group" is a generally useful concept, especially if it's optional, as this PR addresses (the group maps to a "stream" in the dataflow case, but could be applied to any arbitrary app-grouping requirement). We can extend the group-based functionality in the SPI to support group-level status checks as well group-level lifecycle management. We have already identified other areas where this could be useful (e.g. composite REST services), and if we don't support group at the SPI level, then each caller that deals with a group of apps would be required to duplicate much of that status/lifecycle logic.

I agree it would be nice to have the ability to undeploy apps that were not deployed via the SPI, but currently the contract is such that the return value from `deploy` is the ID to use for `status`/`undeploy`. See: https://github.com/spring-cloud/spring-cloud-deployer/blob/master/spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/app/AppDeployer.java#L45

For some runtime environments, there might not be any other alternative to ensure uniqueness (and impractical to put that burden on the caller). I'll let @jvalkeal and @trisberg weigh in for the Yarn, Mesos, and Kubernetes cases.... perhaps we can consider abstracting the namespace(k8s)/domain(CF) features to help in this regard. FWIW I initially resisted the idea of a `DeploymentIdRepository`. At the very least, we should add a method for discovering already deployed apps based on some criteria... that would then populate the repository (and we would need a method to list all known apps).

**Comment by dsyer** [comment]
If "group" is part of the SPI then it should surface in a more strongly typed way than an entry in an optional map. The change in this PR at least would let me decide as a user of the API, that I don't know or care about groups. Can we merge it?

**Comment by markfisher** [comment]
@dsyer I agree that we should either make "group" a first-class part of the API or drop it. Personally, I do think we should extend the API to provide that support for status/lifecycle-management for groups of apps since it will otherwise lead to duplication for any consumers of this API that do need that level of functionality (and I think we already expect that in at least a few cases... data flow already, spinnaker AFAIK, and also for creating composite REST services). That discussion need not hold up this PR though, and I will merge as-is for now.

**Comment by dsyer** [comment]
> I do think we should extend the API

`SmartAppDeployer`?

Can we change the name to "ApplicationDeployer" anyway?

**Comment by markfisher** [comment]
I'd rather avoid the term "Smart"... maybe something like `ApplicationGroupDeployer` if we don't want to add the group-related methods to the core interface. I'm not that opposed to "App" but if it's the consensus, I'm okay with the name change to full "Application", but we would need to do that in many classes (as well as many implementations).

**Comment by markfisher** [comment]
rebased, squashed and merged as https://github.com/spring-cloud/spring-cloud-deployer-cloudfoundry/commit/9f78125ebbc27d8c4ee74b6d04843a5688692e4f

**Pull request closed without merging**